README.md
# Keirin

![Picture of a female track cyclist taking part in a Keirin race. The background is blurred due to her speed.](https://static.wixstatic.com/media/3466eb_ce384529b4fc4cbc913b6cef5d18de37~mv2.jpg/v1/crop/x_1379,y_0,w_3863,h_3854/fill/w_339,h_338,al_c,q_80,usm_0.66_1.00_0.01/UCI_Track_World_Championships_2020-02-29.webp)

Keirin is a microbenchmarking library for Clojure. Microbenchmarking is tricky on the JVM platform as background tasks such as garbage collection can
distort benchmarks. Keirin takes a five pronged approach to the problem.

1) Taking steps to avoid unpredictable events:
 * warming up the JIT to ensure code is compiled 
 * forcing a GC to reduce the chance of a GC during a test run

2) Detecting background events (GC, code compilation, class loading and unloading) when they occur during a trial run and discarding this run.

3) Reducing the impact of timing inaccuracies: 
 * On each timed run, the benchmarked function is run repeatedly for a minimum time period and the mean time calculated. This reduces the inaccuracies in trying to time a function that runs for a very short duration.

4) Using statistical techniques to minimize the impact of outliers:
 * Multiple timed runs are taken and the median of these (rather than the mean) is used for the final result. The median is more robust than the mean in the event of outliers.
 * MAD (["Median absolute deviation"](https://en.wikipedia.org/wiki/Median_absolute_deviation)) is provided as a measure of the variance, rather than the
 sample standard deviation. Again MAD is more robust against outliers than the sample standard deviation.

5) If required, estimating the timing overhead unavoidably included in the times for the benchmarked function.


## Installation
Keirin is available from Clojars.

[![Clojars Project](https://img.shields.io/clojars/v/rachbowyer/keirin.svg)](https://clojars.org/rachbowyer/keirin)

Keirin works most optimally if garbage collections are logged to a file. This can be done by adding the following 
JVM option -Xlog:gc:gc.out. The parser for JVM properties is pretty naive and setting -Xlog:gc options is 
likely to break it.


## Usage
To make the Keirin functions available type:

     (require '[keirin.core :as k])


And to benchmark some code call **k/bench** e.g. 

    (k/bench (Thread/sleep 100))
    ;; Time taken (median) 102.60 ms
    ;; Mean absolute deviation (MADS)  533.19 µs
    
        
**k/bench** takes five optional arguments: 
 * :calc-timing-overhead calculates the overhead due to the timing loop. See the section *Execution overhead* below for more details.
 * :min-execution-time is the minimum time in milliseconds for which the benchmarked function will be run. Defaults to 300ms. If there are problems with garbage collection then this parameter should be reduced.
 * :reporting. Defaults to :display-humane which shows the execution time and the mean absolute deviation. Alternatively :reporting can be :underlying-results, which returns more detailed information as a Clojure hashmap. This can be useful if the results need to be processed by Clojure code e.g. to generate a graph
 * :timed-runs is the number of timed runs required. Defaults to 30.
 * :verbose a boolean, which provides extended output. 

For example: 

    (k/bench (Thread/sleep 500) :num-timed-runs 5 :min-execution-time 1000 :reporting :underlying-results :verbose true)
    
    
If :reporting is set to :underlying-results then the following output is generated:

    (k/bench (Thread/sleep 100) :reporting :underlying-results)
    ;; {:timed-runs 30,
    ;;  :gc-failures 0,
    ;;  :compilation-failures 0,
    ;;  :class-loading-failures 0,
    ;;  :mean 102.70095080000002
    ;;  :median 102.37582549999999,
    ;;  :MAD 0.3478315000000052,
    ;;  :std 1.0023317249709214,
    ;;  :final-gc-time 0.108144995,
    ;;  :data [{:time-taken 104.54769399999999, :result-hash-code nil}
    ;;         {:time-taken 101.031879, :result-hash-code nil}
    ;;          ...]}
 
 * :timed-runs is the number of timed runs
 * :gc-failures, :compilation-failures and :class-loading-failures are the number of times that a run failed due to GC, compilation, class loading/unloading respectively
 * :mean is the mean time that **Thread/sleep** ran taken across the trials in the chosen set in milliseconds
 * :median is the median time that **Thread/sleep** ran taken across the trials in the chosen set in milliseconds
 * :MAD median absolute deviation of the time taken across the successful trials in the chosen set of runs
 * :std is the sample standard deviation of the time taken across the successful trials in the chosen set of runs
 * :final-gc-time is the time in milliseconds of a GC run at the end of the set of runs. 
 * :data contains the time taken by each timed run and the hash code of the result of the function


The **k/quick-bench** macro is also provided. This behaves in the same way as the **bench** macro, but only performs 7 rather than 30 timed runs.  

    (k/quick-bench (Thread/sleep 100))
    ;; Time taken (median) 102.85 ms
    ;; Mean absolute deviation (MADS)  426.87 µs


See https://github.com/rachbowyer/keirin-demo for an extended demo using Keirin and the Analemma library to graph the performance
of **vec** as the number of elements in the vector increases.


## Execution overhead

To minimise the volatility of results, and also to increase accuracy for very fast functions, the function under test is run repeatedly for at least 300ms and the total duration of the run is timed. This means unfortunately that there is some overhead added to the results as parts of the timing loop 
are timed as well. 

Also on most (if not all computers) System/nanoTime, which Keirin uses to time the function is not nano second precise and also introduces its own overhead. 

For example:

    (bench (System/nanoTime))
    ;; Time taken (median) 38.38 ns
    ;; Mean absolute deviation (MADS)  0.57 ns

and

     (let [timings (doall (for [_ (range 1000)]
                       (let [start (System/nanoTime)
                             end   (System/nanoTime)]
                             (- end start))))]
  
       (println "Range " (apply min timings) " to " (apply max timings) " nanos"))
     ;; Range  50  to  374  nanos

The overhead included in the number that Keirin calculates depends on the number of times the function under test is executed within a timed run and therefore differs for different functions and also for different values for :min-execution-time.

In my tests (YMMV) the amount of the overhead varies from 2 ns to 2 µs with the overhead increasing with slower functions. The maximum overhead I have seen is around 15% for functions that are running in around 25 nanos, dropping to less than 0.001% as the function execution time increases.

Keirin does not attempt to automatically compensate for this overhead, but if the execution overhead is a concern, pass in the flag :calc-timing-overhead and Keirin will provide an estimate of the overhead.


    (bench (vector-of :int 1 2 3 4) :calc-timing-overhead true)
    ;; Time taken (median) 25.90 ns
    ;; Mean absolute deviation (MADS)  0.62 ns
    ;; Timing loop overhead  14.17% (3.67 ns)
    
    (bench (Thread/sleep 100) :calc-timing-overhead true)
    ;; Time taken (median) 103.45 ms
    ;; Mean absolute deviation (MADS)  167.87 µs
    ;; Timing loop overhead  0.00% (1.96 µs)


## Benchmarking best practice

 1) Run on a machine with multiple cores, ample memory, which does not have a GUI and is quiet. This minimizes the chance of the OS or JVM stealing CPU from the benchmarking. For example use an unused server class machine. Machines in the Cloud are not considered reliable for benchmarking purposes.
   
 2) Compile down to a jar and run the code as a jar, rather than run from the REPL as the REPL introduces noise.

 3) Configure the JVM correctly. For example the following options are useful:
  * "-server" - enables server side JVM optimisations 
  * "-Xbatch" - disable background compilation
  * "-Xlog:gc:gc.out" - logs GC to a file where Keirin can read them
  * "-XX:-TieredCompilation" - disables tiered compilation.
  * "-Xmx" - sets the maximum heap size. Set as large as possible for your machine as this reduces the chance of garbage collections.
  * "-Xms" - sets the minimum heap size. Set to the same value as Xmx as this reduces the chance of garbage collections.


## Rationale for Keirin

See this [article](https://www.bowyer.info/clojure-keirin) on my website.


## License for Keirin

Distributed under the Eclipse Public License version 1.0.


<hr>


<small>Picture of a cyclist courtesy of Tim Rademacher. Used under the Creative Commons [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/) licence.</small>




CHANGELOG.md
# Change Log

## [1.1.2] - 2021-04-15

### Changed
- Adjust timing loop so it is run twice. Times from the first run are disregarded. 
  The idea is to use the first run to trigger any runtime optimisations prior to 
  the second timed run.


## [1.1.1] - 2021-03-10

### Changed
- Now runs on Java 11.


test/keirin/core_test.clj
;; Copyright ©️ Rachel Bowyer 2017, 2021. All rights reserved.
;;
;; This program and the accompanying materials
;; are made available under the terms of the Eclipse Public License v1.0
;; which accompanies this distribution, and is available at
;; http://www.eclipse.org/legal/epl-v10.html

(ns keirin.core_test
  (:require [keirin.core :as kc]
            [clojure.test :refer :all]))

(def ^:private sq #'kc/sq)
(def ^:private abs #'kc/abs)
(def ^:private mean #'kc/mean)
(def ^:private sample-variance #'kc/sample-variance)
(def ^:private sample-standard-deviation #'kc/sample-standard-deviation)
(def ^:private median #'kc/median)
(def ^:private MAD #'kc/MAD)
(def ^:private parse-gc-file-name #'kc/parse-gc-file-name)

(deftest sq-test
  (are [x r] (= r (sq x))
    0   0
    2   4
    -1  1))

(deftest abs-test
  (are [x r] (= r (abs x))
     0    0.0
     -1   1.0
     -1.5 1.5
     2    2.0))

(deftest mean-test
  (are [x r] (= r (mean x))
     [2]          2.0
     [4 2 3 1]    2.5
     [2 4 6 8]    5.0))

(deftest sample-variance-test
  (are [x r] (= r (sample-variance x))
     [1]        0.0
     [1 2]      0.5
     [1 2 3]    1.0
     [1 2 3 4]  (/ 5 3.0)))

(deftest sample-standard-deviation-test
  (are [x r] (= r (sample-standard-deviation x))
     [1]        0.0
     [1 2 3]    1.0
     [2 4 6]    2.0))

(deftest median-test
  (are [x r] (= r (median x))
      [1]         1.0
      [1 2]       1.5
      [1 3 2]     2.0
      [10 2 6 4]  5.0))

(deftest MAD-test
  (are [x r] (= r (MAD x))
      [1 1 1 1]   0.0 ; All elements are 0 from the central point
      [1 2 3 4]   1.0
      [1 2 3 11]  1.0 ; Large outlier does not disturb median or MAD
      [2 8 4 6]   2.0))

(deftest parse-gc-file-name-test
  (is (= "gc.out" (parse-gc-file-name "-Xlog:gc:gc.out"))))

project.clj
;; Copyright ©️ Rachel Bowyer 2016, 2017, 2021. All rights reserved.
;;
;; This program and the accompanying materials
;; are made available under the terms of the Eclipse Public License v1.0
;; which accompanies this distribution, and is available at
;; http://www.eclipse.org/legal/epl-v10.html

(defproject rachbowyer/keirin "1.1.2"
  :description "Microbenchmarking library for Clojure"
  :url "https://github.com/rachbowyer/keirin"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}

  :dependencies [[org.clojure/clojure "1.10.1"]
                 [org.clojure/tools.logging "0.3.1"]]

  :target-path "target/%s"

  :profiles {:dev {:global-vars {*warn-on-reflection* true *assert* true}
                   :jvm-opts ^:replace ["-Dclojure.compiler.direct-linking=true" "-server"
                                        "-Xlog:gc:gc.out" "-XX:-TieredCompilation" "-Xbatch"]}

             :uberjar {:jvm-opts ^:replace ["-Dclojure.compiler.direct-linking=true"]
                       :aot :all}})


src/keirin/core.clj
;; Copyright ©️ Rachel Bowyer 2016, 2017, 2021. All rights reserved.
;;
;; This program and the accompanying materials
;; are made available under the terms of the Eclipse Public License v1.0
;; which accompanies this distribution, and is available at
;; http://www.eclipse.org/legal/epl-v10.html
;;
;; The functions "heap-used" and "force-gc" are adapted from functions
;; that are marked as "Copyright (c) Hugo Duncan" and distributed as part
;; of the Criterium library (https://github.com/hugoduncan/criterium).
;; These functions are used under the terms of the EPL v 1.0.
;;
;; Inspired by the article
;; http://www.ibm.com/developerworks/java/library/j-benchmark1/index.html
;; by Brent Boyer, the Criterium bench marking library for Clojure (see above)
;; and
;; http://www.javaworld.com/article/2077496/testing-debugging/java-tip-130--do-you-know-your-data-size-.html


(ns keirin.core
  (:require
    [clojure.java.io :as io]
    [clojure.pprint :as pprint]
    [clojure.tools.logging :as log])
  (:import
    [java.lang.management ManagementFactory]))


;;;
;;; Default parameters
;;;

(def ^:private warm-up-executions 10)
(def ^:private warm-up-time
  "seconds"
  10)

(def ^:private default-num-timed-runs 30)
(def ^:private default-num-timed-runs-quick 7)
(def ^:private default-min-execution-time
  "millis"
  200)

(def ^:private max-gc-failures 10)
(def ^:private max-compilation-failures 10)
(def ^:private max-class-loading-failures 5)
(def ^:private max-gc-attempts 100)
(def ^:private default-reporting
  "other option is :underlying-results"
  :display-humane)

(def ^:private min-execution-time-threshold
  "%age of the min execution time that must be achieved"
  0.75)


;;; Memory utilities

(defn heap-used
  "Report a (inconsistent) snapshot of the heap memory used."
  []
  (let [runtime (Runtime/getRuntime)]
    (- (.totalMemory runtime) (.freeMemory runtime))))

(defn force-gc
  "Force garbage collection and finalizers so that execution time associated
   with this is not incurred later. Up to max-attempts are made."
  ([] (force-gc max-gc-attempts))
  ([max-attempts]
   (loop [memory-used (heap-used)
          attempts 0]
     (System/runFinalization)
     (System/gc)
     (let [new-memory-used (heap-used)]
       (if (and (or (pos? (.. ManagementFactory
                              getMemoryMXBean
                              getObjectPendingFinalizationCount))
                    (> memory-used new-memory-used))
                (< attempts max-attempts))
         (recur new-memory-used (inc attempts)))))))


;;;; Stats functions

(defn- sq [x]
  (* x x))

(defn abs
  ^Double [^Double x]
  (double (Math/abs x)))

(defn- mean
  ^Double [data]
  {:pre [(seq data)]}
  (double (/ (reduce + data) (count data))))

(defn- sample-variance
  "Find the unbiased sample variance"
  ^Double [data]
  {:pre [(seq data)]}
  (let [mean (double (mean data))
        num (count data)]
    (if (> num 1)
      (double (/ (transduce (map #(sq (- % mean))) + data)
                 (dec num)))
      0.0)))

(defn- sample-standard-deviation
  "Sample standard deviation - based on the unbiased sample variance
   although apparantly ssd is still has some bias"
  [data]
  {:pre [(seq data)]}
  (some-> data sample-variance Math/sqrt))

(defn- median
  "Finds the median of a collection of numbers"
  ^Double [data]
  {:pre [(seq data)]}
  (let [sorted-data (sort data)
        len-data (count data)
        mid-point (quot len-data 2)]
    (if (odd? len-data)
      (double (nth sorted-data mid-point))
      (/ (+ (nth sorted-data (dec mid-point))
            (nth sorted-data mid-point)) 2.0))))

(defn- MAD
  "Finds the median absolute deviation"
  ^Double [data]
  {:pre [(seq data)]}
  (let [central-point (median data)]
    (median (map (fn [e] (abs (- e central-point)))
                 data))))


;;;; Bench marking utilities - speed

(defn- parse-gc-file-name [jvm-arg]
  (let [[_ gc-file] (re-find #"^-Xlog:gc:(.*?)$" jvm-arg)]
    gc-file))

(defn- get-gc-file-name []
  (let [jvm-options (.getInputArguments (ManagementFactory/getRuntimeMXBean))]
        (some parse-gc-file-name jvm-options)))

(defn- read-gc-file [gc-file-name]
  (if (.exists (io/as-file gc-file-name))
    (slurp gc-file-name)
    ""))


;;; We need an ultra fast place to store the result of running the "payload"
;;; function
;;; as storing it is included in the timing loop

(defprotocol TimedRunResultAccessor
  (set-result! [_ v] )
  (get-hash-code [_]))

(deftype TimedRunResultType [^{:unsynchronized-mutable true :tag Object} v]
  TimedRunResultAccessor
  (set-result! [_ value] (set! v value))
  (get-hash-code [_] (some-> v .hashCode)))

(def timed-run-result (TimedRunResultType. nil))

(defn execute-many [f n]
  (if (= n 1)
    (let [start (System/nanoTime)
          ^Object result (f)]

      (let [end (System/nanoTime)]
        {:time-taken       (/ (- end start) 1000000.0)
         :result-hash-code (some-> result .hashCode)}))

    ;; This is the main execution loop
    ;; Run it twice in succession - to allow JIT to do maximum
    ;; optimisations
    (loop [second-time false]
      (let [start (System/nanoTime)]
        (loop [i (long n)]
          (when (pos? i)
            (set-result! timed-run-result (f))
            (recur (unchecked-dec i))))

        (let [end (System/nanoTime)]
          (if-not second-time
            (recur true)
            {:time-taken       (/ (- end start) 1000000.0 n)
             :result-hash-code (get-hash-code timed-run-result)}))))))

(defn- bench-one [payload gc-file-name n & {:keys [no-force-gc verbose]}]
  (when verbose
    (println "Requesting GC..."))
  (when-not no-force-gc
    (force-gc))
  (Thread/sleep 300) ;; Let any GC finish
  (when verbose
    (println "GC complete"))

  (let [compilation-bean
        (ManagementFactory/getCompilationMXBean)

        compilation-time-supported
        (.isCompilationTimeMonitoringSupported compilation-bean)

        class-loading-bean
        (ManagementFactory/getClassLoadingMXBean)

        class-loaded-before
        (.getTotalLoadedClassCount class-loading-bean)

        class-unloaded-before
        (.getUnloadedClassCount class-loading-bean)

        compilation-time-before
        (and compilation-time-supported
             (.getTotalCompilationTime compilation-bean))

        gc-file-before
        (some-> gc-file-name read-gc-file)

        {:keys [result-hash-code time-taken]}
        (execute-many payload n)

        compilation-time-after
        (and compilation-time-supported
             (.getTotalCompilationTime compilation-bean))

        class-loaded-after
        (.getTotalLoadedClassCount class-loading-bean)

        class-unloaded-after
        (.getUnloadedClassCount class-loading-bean)

        _
        (Thread/sleep 300) ;; Let GC file catch up

        gc-file-end
        (some-> gc-file-name read-gc-file)]

      {:gc-occurred
       (not= gc-file-before gc-file-end)

       :compilation-occurred
       (not= compilation-time-before compilation-time-after)

       :class-loading-occurred
       (or (not= class-loaded-before class-loaded-after)
           (not= class-unloaded-before class-unloaded-after))

       :time-taken
       time-taken

       :result-hash-code
       result-hash-code}))

(defn- iterate-bench
  [payload gc-file-name n & {:keys [verbose num-timed-runs] :as options}]
  (loop [gc-failure-count             0
         compilation-failure-count    0
         class-loading-failure-count  0
         iters                        0
         times                        []]

    (when verbose
      (println "Starting timed run..."))
    (let [{:keys [gc-occurred compilation-occurred class-loading-occurred
                  time-taken result-hash-code]}
          (apply bench-one payload gc-file-name n (-> options seq flatten))

          failure
          (or gc-occurred compilation-occurred class-loading-occurred)

          new-gc-failure-count
          (cond-> gc-failure-count gc-occurred inc)

          new-compilation-failure-count
          (cond-> compilation-failure-count compilation-occurred inc)

          new-class-loading-failure-count
          (cond-> class-loading-failure-count class-loading-occurred inc)

          new-iters
          (cond-> iters (not failure) inc)

          new-times
          (cond-> times
                  (not failure)
                  (conj {:time-taken       time-taken
                         :result-hash-code result-hash-code}))]

      (when verbose
        (println "Run complete " time-taken "ms. GC occurred " gc-occurred
                 ". Compilation occurred " compilation-occurred
                 ". Class loading occurred " class-loading-occurred
                 ". Result hash code " result-hash-code ". "))

      (if (or (>= new-gc-failure-count max-gc-failures)
              (>= new-compilation-failure-count max-compilation-failures)
              (>= new-class-loading-failure-count max-class-loading-failures)
              (>= new-iters num-timed-runs))
        {:data                    new-times
         :gc-failures             new-gc-failure-count
         :compilation-failures    new-compilation-failure-count
         :class-loading-failures  new-class-loading-failure-count}
        (recur new-gc-failure-count new-compilation-failure-count
               new-class-loading-failure-count new-iters new-times)))))

(defn- warm-up!
  "Does warm-up-executions of benchmarked code and runs for at least
   warm-up-time seconds. Need to run for at least 10secs to ensure that the
   JVM gets the message and compiles code, rather than just interpret it."
  [payload verbose]

  (when verbose
    (println "Warming up JVM..."))

  (let [start-time
        (System/nanoTime)

        [iterations time]
        (loop [i 0]
          (execute-many payload 1)
          (let [end-time (System/nanoTime)
                execution-time (/ (- end-time start-time) 1e9)]
            (if (and (>= i warm-up-executions)
                     (>= execution-time warm-up-time))
                [i execution-time]
                (recur (inc i)))))]
  (when verbose
    (println "JVM warmed up. Iterations " iterations
             ". Time taken " time "."))))

(defn- bench-one-set
  [payload gc-file-name n & {:keys [verbose num-timed-runs
                                    min-execution-time no-force-gc]
                             :or {num-timed-runs default-num-timed-runs}
                             :as options}]

  (let [{:keys [data gc-failures compilation-failures class-loading-failures]}
        (apply iterate-bench payload gc-file-name n
               (-> options
                   (merge {:num-timed-runs num-timed-runs})
                   seq
                   flatten))]

    ;; Do final GC and measure time - to ensure there is not a big GC mess
    (when verbose
      (println "Starting final GC..."))

    (let [start-time (System/nanoTime)]
      (when-not no-force-gc
        (force-gc))
      (let [end-time (System/nanoTime)]

        (when verbose
          (println "Final GC complete"))

        (when verbose
          (pprint/pprint data))

        (when (or (empty? data)
                  (not= (count data) num-timed-runs))
          (let [error-msg
                (str "Unable to complete " num-timed-runs " timed runs. "
                     "GC failures " gc-failures ". "
                     "Compilation failures " compilation-failures ". "
                     "Class loading failures " class-loading-failures ". ")]

            (throw (Exception. error-msg))))
        (let [times   (map :time-taken data)
              result  {:num-executions         n
                       :timed-runs             (count data)
                       :gc-failures            gc-failures
                       :compilation-failures   compilation-failures
                       :class-loading-failures class-loading-failures
                       :mean                   (double (mean times))
                       :median                 (median times)
                       :std                    (sample-standard-deviation times)
                       :MAD                    (MAD times)
                       :final-gc-time          (/ (- end-time start-time)
                                                  1000000000.0)
                       :data data}
              execution-time  (* (apply min times) n)]

          (when (and min-execution-time
                     (< execution-time (* min-execution-time
                                          min-execution-time-threshold)))
            (let [error-msg
                  (str "Did not run payload function for long enough to get"
                        " accurate results. Ran payload function for only "
                        execution-time "ms, when the minimum  required "
                       "execution time is " min-execution-time "ms.")]
              (throw (Exception. error-msg))))
          result)))))

(defn- humane-time
  "t is time in milliseconds"
  [t]
  (let [time-seconds (/ t 1000.0)]
    (cond
      (and (< time-seconds 1) (>= time-seconds 0.001))
      (format "%.2f ms" (* time-seconds 1000.0))

      (and (< time-seconds 0.001) (>= time-seconds 0.000001))
      (format "%.2f µs" (* time-seconds 1000000.0))

      (< time-seconds 0.000001)
      (format "%.2f ns" (* time-seconds 1000000000.0))

      :else (format "%.2f seconds" time-seconds))))


(defn- display-humane
  [{:keys [timing-overhead median MAD]:as _results}]
  (println "Time taken (median)" (humane-time median))
  (println "Mean absolute deviation (MADS) " (humane-time MAD))
  (when timing-overhead
    (let [as-percentage (* 100 (/ timing-overhead median))]
     (println "Timing loop overhead "
              (str (format "%3.2f" as-percentage)
                   "% ("
                   (humane-time timing-overhead)
                   ")")))))

(defn- report-results [results reporting]
  (condp = reporting
    :display-humane     (display-humane results)
    :underlying-results results
    results))

(defn- estimate-num-executions
  "Estimate the number of times to execute the payload function in order to
   achieve the required minimum execution time.

   Timing an empty block with two successive calls to System/nanoTime on a
   2013 Macbook Pro can take as long as 300 nanos in the REPL and 150 nanos in
   compiled code. This introduces a considerable error for a function that only
   takes 10 nanos to run, but the error reduces as the number of executions
   increases.

   However, on the other hand slower functions may allocate a considerable
   amount of memory. If they are run too often then GCs may occur in the timing
   run invalidating.

   The strategy is to start with an initial estimate. If the minimum required
   execution time is not reached, then re-estimate based on the time executing
   this one function. The additional executions will reduce the timing error
   and likely lead to the payload function appearing to speed up. Therefore the
   new estimate will probably alsobe too low but more accurate. Eventually it
   will stabilize giving an accurate estimate.

   Critical that we get fairly accurate results - one slow run could throw the
   execution count right off - and impact results. So use min of 3 for each
   time"
  [payload gc-file-name min-execution-time verbose]
  (loop [n 1]
    (when verbose
      (println "Trying " n " executions"))

    (let [{:keys [data gc-failures compilation-failures
                  class-loading-failures] :as result}
          (iterate-bench payload gc-file-name n
                         :num-timed-runs 3
                         :verbose verbose)

          _ (when-not (= 3 (count data))
              (let [error-msg
                    (str "Unable to get clean runs estimating number of "
                         "executions. GC failures: " gc-failures ". "
                         "Compilation failures: " compilation-failures "."
                         "Class loading failures: " class-loading-failures
                         ". ")]
                (throw (Exception. error-msg))))

          time-taken
          (->> data (map :time-taken) (apply min))

          time-taken
          (max time-taken 0e-6) ; In case we get back 0 for a fast function

          total-time-taken
          (* time-taken n)]

      (if (> total-time-taken min-execution-time)
        n
        (let [new-n (inc (int (quot min-execution-time time-taken)))
              ;; failsafe to ensure monotonically increasing
              new-n (max new-n (inc n))] ;
          (recur new-n))))))

(defn- calculate-timing-overhead
  [n gc-file-name & {:keys [verbose] :as options}]
  (let [one   (int 1)
        dummy (fn [] one)]

    (when verbose
      (println "Beginning to calculate timing loop overhead..."))

    (warm-up! dummy verbose)
    (force-gc)

    (let [new-options
          (-> options
              (merge {:min-execution-time nil :no-force-gc true})
              seq
              flatten)

          results
          (apply bench-one-set dummy gc-file-name n new-options)

          overhead
          (:median results)]
      (when verbose
        (println "Timing loop overhead " (humane-time overhead)))
      overhead)))

(defn bench*
  [payload & {:keys [verbose min-execution-time reporting calc-timing-overhead]
              :or {min-execution-time default-min-execution-time
                   reporting default-reporting}
              :as options}]
  (let [gc-file-name  (get-gc-file-name)]
    (when-not gc-file-name
      (log/warn "JVM option: -Xlog:gc is not set. Keirin is unable to detect "
                "if a GC occurs during a timed run"))

    (warm-up! payload verbose)

    ;; Calculate number of times to run the payload function
    (let [num-executions
          (estimate-num-executions payload gc-file-name
                                   min-execution-time verbose)

          _
          (when verbose
            (println "Estimated number of executions " num-executions))

          new-options
          (-> options
              (merge {:min-execution-time min-execution-time})
              seq
              flatten)

          results
          (apply bench-one-set payload gc-file-name num-executions new-options)

          timing-overhead
          (when calc-timing-overhead
            (apply calculate-timing-overhead num-executions
                   gc-file-name new-options))

          results-with-overhead
          (cond-> results
                  timing-overhead
                  (assoc :timing-overhead timing-overhead))]

      (report-results results-with-overhead reporting))))

(defmacro bench [payload & options]
  `(bench* (fn [] ~payload) ~@options))

(defn quick-bench*
  [payload & {:keys [num-timed-runs]
              :or {num-timed-runs default-num-timed-runs-quick}
              :as options}]
  (apply bench* payload
         (-> options (merge {:num-timed-runs num-timed-runs}) seq flatten)))

(defmacro quick-bench [payload & options]
  `(quick-bench* (fn [] ~payload) ~@options))